### PR TITLE
feat: avoid pulling unnecessary columns when querying append mode table

### LINKED ISSUE #1302: Wasteful to pull all primary key columns in query of `append` mode
### Describe This Problem

We pull `projected + pk` columns when querying in both `append` and `overwrite` mode.
However pulling  `pk` columns is totally unnecessary in `append` mode.

### Proposal

Just pulling the `projected` columns in query of `append` mode.

The record batch reading steps will be divided to following threes:
+ Pulling the `ArrowRecordBatch`.
+ Converting it to `FetchingRecordBatch`, somethings like filling not exist column a null/default value will be done here.
`FetchingRecordBatch` will be used in `ChainIterator`/ `MergeIterator`, and it may include not only the projected columns but the primary key columns for dedupping or else.
+ Prune to `RecordBatch`, as saying above, `FetchingRecordBatch` can include not only projected columns, prune the non-projecteds here.

Main changes:
+ Don't pass the `ProjectedSchema` including too many informations to where building the inner `RecordBatchStream` in `ChainIterator`/ `MergeIterator`. Instead, refactor the `RowProjector` to just include the needed informations and pass it to(mainly `ScanRequest` and `SstReadOptions`).
+ Refactor `RecordBatchWithKey` to `FetchingRecordBatch`, the main difference is that `FetchingRecordBatch` can include `primary_keys_indexes` or not. Actually, `FetchingRecordBatch` should not include `primary_keys_indexes` anymore, but it is hard to remove it completely, so maybe we can delay it to later prs.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Rachelint**
## Rationale
Closes #1302 

The pulling arrow record batches are ensured to include primary key columns, however the pulled primary key columns are unused for append mode tables' queries.
I refactor the whole record batches pulling path in this pr for readability and enhancement for avoiding pulling primary key columns even they are unused.

## Detailed Changes
+ Refactor `RowProjector` to `RecordFetchingContext` holding just the needed information, and pass it to `ScanRequest` & `SstReadOptions` rather than the too heavy `ProjectedSchema`.
+ Refactor `RecordBatchWithKey` to `FetchingRecordBatch` which holds the primary indexes on demand.

## Test Plan
Test by exist and new added tests.

### CODE REVIEW

File: analytic_engine/src/instance/read.rs [hunk 1]
```diff
@@ -189,15 +197,15 @@ impl Instance {
...
+        sst_read_options_builder: SstReadOptionsBuilder,
     ) -> Result<Vec<DedupIterator<MergeIterator>>> {
         // Current visible sequence
         let sequence = table_data.last_sequence();
         let time_range = request.predicate.time_range();
         let version = table_data.current_version();
         let read_views = self.partition_ssts_and_memtables(time_range, version, table_options);
         let iter_options = self.make_iter_options(table_options.num_rows_per_row_group);
-
+        // generate builder
```
Comment (ShiKaiWi) [inline analytic_engine/src/instance/read.rs:208]: Remove this line?

File: common_types/src/projected_schema.rs [hunk 2]
```diff
@@ -61,33 +61,160 @@ pub enum Error {
 pub type Result<T> = std::result::Result<T, Error>;
 
 #[derive(Debug, Clone)]
-pub struct RowProjector {
-    schema_with_key: RecordSchemaWithKey,
+pub struct RecordFetchingContext {
```
Comment (ShiKaiWi) [inline common_types/src/projected_schema.rs:64]: `RowProjector` seems a better name.

File: common_types/src/projected_schema.rs [hunk 3]
```diff
@@ -61,33 +61,160 @@ pub enum Error {
...
+    ///
+    /// For example:
+    ///   source column indexes: 0,1,2,3,4
+    ///   data fetching indexes in source: 2,1,3
+    ///
+    /// We can see, only columns:[1,2,3] in source is needed,
+    /// and their indexes in pulled projected record bath are: [0,1,2].
+    ///
+    /// So the stored data fetching indexes in projected source are: [1,0,2].
+    fetching_projected_source_column_indexes: Vec<Option<usize>>,
```
Comment (ShiKaiWi) [inline common_types/src/projected_schema.rs:95]: Propose a better name `projection_on_arrow_record_batch`?

**Review by jiacai2050 — APPROVED** [review]

**Review by tanruixiang — APPROVED** [review]
LGTM. merge after ci pass

**Pull request merged**